### PR TITLE
WIP initial commit to fix #5

### LINKED ISSUE #5: Implement `AttestationManager`
All the offers received by the `BoardManager`, independently of the communications channels (BOLT4 onions, BOLT8 noise channel or NIP 01 websocket) and sources (clients or peers).

The `AnchorManager` (in `anchormanager.rs`) should accept serialized offer payload from the `BoardManager`, pack them in an anchor and at periodic intervals (e.g every X blocks or clock ticks), counter-sign them with the board public key and submit for validation to the selected Mainstay server.

The anchor format should be some metadata (version, time, previous anchor SHA256 hash if proof-of-immutable-sequence is opted-in) and a byte payload.

The Mainstay protocol is documented here: https://commerceblock.readthedocs.io/en/latest/mainstay-background/index.html and the original paper is available here: https://cloudflare-ipfs.com/ipns/ipfs.commerceblock.com/commerceblock-whitepaper-mainstay.pdf

### PR DESCRIPTION
**Pull request by twshelton**
1. this module will need to accept a serialized offer payload.  I have added some basics to allow for this but need to understand what type of data will be provided.  Will this be an array of u64? Array of strings/binaries? Other? 
2. Will this module be responsible for the timer used to determine when to write to Mainstay server or will that be an external process that uses this module on set intervals to send to Mainstay.
3. Will the Mainstay server be accessed via HTTP API or c-callable wrapper? I need to dig a little deeper here to understand the options based on the desired approach.

### CODE REVIEW

**Comment by ariard** [comment]
> this module will need to accept a serialized offer payload. I have added some basics to allow for this but need to understand what type of data will be provided. Will this be an array of u64? Array of strings/binaries? Other?

I think here we have two-level of payloads serialization:
- internal: between the `AttestationManager` and the different components like the nostr's `ClientHandler` or `ServiceManager`
- external: between the `AttestationManager` p2p-stack and the Mainstay server

Between the internal components, I think we can have Nostr events, that way being able to have something unified to provide for the attestation, a la NIP-03-https://github.com/nostr-protocol/nips/blob/master/03.md

For the external flow, I think you can have something compatible with Mainstay server interface, with metadata necessary for proof-of-immutable-state.

> Will this module be responsible for the timer used to determine when to write to Mainstay server or will that be an external process that uses this module on set intervals to send to Mainstay.

That's a good question, at the very least under the Civkit architecture as described in the whitepaper, you have at least two services:
- the market bulletin boards
- all types of oracles (e.g moderators or "know your peer")

Both of them might have different service-level requirements in terms of time intervals about when to do Mainstay attestations.

> Will the Mainstay server be accessed via HTTP API or c-callable wrapper? I need to dig a little deeper here to understand the options based on the desired approach.

For now, I think we can assume some HTTP API over TCP or Noise-based BOLT8-style connections (more modern cryptography than TLS). Ideally, we could have two interfaces, one network-level and one inter-process to have the `AttestationManager` accessible by `civkit-cli` over a gRPC interface or cap'n proto one (like Core is doing for [multiprocess](https://github.com/bitcoin/bitcoin/pull/10102)).

**Comment by twshelton** [comment]
I have added the ability to create a new AnchorManager and append to its collection of Offers.  I need some assistance with dereferencing the offer before it is appended if this design is acceptable.  Also, I'm assuming that the BoardManager will create a new reference to AnchorManager and provide it's board_pubkey.  This is causing a problem when the BoardManager is created in server.rs as the AnchorManager is created then the BoardManager.  An alternative would be to have the BoardManager pass it's pubkey on each call to AnchorManager.

With this approach, we should also consider what happens to the offers in memory should a panic occur.  Should we consider a database or some other serialization to guard against data loss?

**Comment by ariard** [comment]
First and foremost, could you rebase on top of `8a17a4f` ? Note you have committed `Cargo.lock` in your push that is making a big review diff :)

> An alternative would be to have the BoardManager pass it's pubkey on each call to AnchorManager.

Let’s go for this approach, as that way we have architecture flexibility if services like market board or mainstay services wanna _switch_ their keys in function of the attestation context.

> With this approach, we should also consider what happens to the offers in memory should a panic occur. Should we consider a database or some other serialization to guard against data loss?

Adding a persistent database is a very good remark, though let’s do it in another PR.

For this one, it’s all good if we land `AnchorManager` as a component accepting generic `add_payload()` from the `BoardManager`.

File: src/anchormanager.rs [hunk 1]
```diff
@@ -1,23 +1,69 @@
...
+// use civkit::events;
+// use civkit::events::ClientEvents;
+// use civkit::kindprocessor::KindProcessor;
+
+pub struct AnchorManager {
+    our_service_pubkey: Arc<PublicKey>, 
+    our_service_offers: Arc<SerializedOffers>,     
+}
+
+pub struct SerializedOffers {
```
Comment (ariard) [inline src/anchormanager.rs:28]: Can we call this `SerializedEvent` as the granular data structure we’re aiming to commit for attestations is more-or-less a Nostr event with some additional metadatas ?

File: src/anchormanager.rs [hunk 2]
```diff
@@ -1,23 +1,69 @@
...
+        our_service_offers: Arc::new(offers)
+    }
 	}
 
+  pub fn add_payload(&mut self, serialized_offer: Arc<SerializedOffer>) -> () {
+    let offer = Arc::clone(&serialized_offer);
+    //self.our_service_offers.push(&offer)
+  }
+
 	fn commit_note(&self) {}
```
Comment (ariard) [inline src/anchormanager.rs:62]: You can remove `commit_note` it’s quite the same than `add_payload`.

**Comment by ariard** [comment]
Let me know if you would like more feedback on the git rebase process, here to answer questions.

**Comment by twshelton** [comment]
I'm not sure I completely understand the rebase process as this PR shows 16 commits but only 1 file changed.  Let me know if I need to make further changes.

This PR should provide the baseline functionality.  We still need to incorporate the database to persist events between Mainstay calls.  We also need to add the Mainstay process itself and implement the _timed_ writing of the events.

**Comment by ariard** [comment]
> This PR should provide the baseline functionality. We still need to incorporate the database to persist events between Mainstay calls. We also need to add the Mainstay process itself and implement the timed writing of the events.

Many thanks for the progress here. As discussed on the discord server, finally introduced a refactoring of our gRPC API `civkitservices` against which I think the Mainstay process should be build on, see #45.

I’ll review more this PR, just finishing #45 first.

**Comment by ariard** [comment]
Added the new https://github.com/civkit/civkit-node/pull/61/files with boilerplate code using the new `civkitservice` interface and a list of todos to build a timestamping notarization server.

I think the first code step is to receive the queries from Nostr clients (e.g `civkit-sample` for us) and print them or store them in a new notary local map (`Vec` / `HashMap`). Note there might be a need to add first a new event kind in the downstream library: https://github.com/rust-nostr/nostr

I think the work would benefit from (BIP ?) standardization of notarization attestation, namely the data payload which is generated by the Nostr client and relayed to the notarization server.

**Comment by ariard** [comment]
Closing this PR as `notaryd` is moving the conversation forward in matter of notarization. If any element is forgotten from this conversation here, please raised it and we can re-open it.

**Pull request closed without merging**